_source/_posts/2021-09-12-spring-boot-spa.md

```diff
@@ -0,0 +1,686 @@
+        onAuthRequired: () => {
+            router.push('/login')
+        },
+        onAuthResume: () => {
+            router.push('/login')
+        },
+    })
+    .mount('#app')
+```
+
+You can now run the Vue app.
+
+```bash
+npm run server
+```
+
+## Test the Completed Server and Client
+
+Make sure your resource server is still running. Don't forget that you will need to have sourced the `.okta.env` file in the shell you use to run the Spring Boot app. Here are the commands in case you need to restart the server (which, of course, need to be run from the Spring Boot project's root directory).
+
+```bash
+source .okta.env
+./mvnw spring-boot:run
+```
+
+Open the app in your browser at [http://localhost:8080](http://localhost:8080).
+
+Because the app is set to secure the home page, you will immediately be directed to the Okta login screen. Log in using your Okta credentials. 
+
+{% img blog/spring-boot-spa/okta-sign-in.png alt:"Okta Sign-In Page" width:"800" %}{: .center-image }
+
+When you enter your credentials, the Okta Sign-In Widget and the Vue app will follow the OAuth 2. 0 Authorization Code flow. In this flow, the client sends the login credentials to the Okta auth server and, if authentication is successful, receives an authorization code. The client app calls the Okta token endpoint, trading this authorization code for a JWT (JSON Web Token). The standard OAuth 2.0 authorization code flow requires that the application send the client secret along with the code to the token endpoint. 
+
+However, in the context of a client-side application, putting the client secret in public, browser code would be a major security violation. Instead, Okta uses PKCE (Proof Key for Code Exchange). In this modified flow, the client generates a one-time key that is sent with the request and is associated with the authorized JWT. This is used to ensure that only the client that requested the JWT can use it.
+
+There's a great, in-depth article on the PKCE flow [on the Okta Blog by Micah Silverman](https://developer.okta.com/blog/2019/08/22/okta-authjs-pkce).
+
+Once you've logged in you'll see the simple application screen below.
+
+{% img blog/spring-boot-spa/success-screen.png alt:"Our simple application" width:"800" %}{: .center-image }
+
+Notice that it has pulled your name and email from your authentication information. It has also made an HTTP request to the Spring Boot REST server using the JWT to determine the caffeine level of your developer (the `/howcaffeinatedami` endpoint). Don't worry! Although the image above is reporting your developer's caffeine level is "not at all," never fear. I can promise you this is incorrect and simply a random response merely for mildly humorous effect.
+
+You can also go to the `profile` page at http://localhost:8080/profile, which will show you all of the claims in the authenticated token.
+
+{% img blog/spring-boot-spa/profile-page.png alt:"Profile page" width:"800" %}{: .center-image }
+
+## Conclusion
```

Comment:
nit: Add a more descriptive title, "Learn More About Building Vue and Spring Boot Apps", or something